src/compiler/checker.ts

```diff
@@ -6653,10 +6653,98 @@ namespace ts {
 
             return type;
 
+            function narrowTypeByValueExpression(type: Type, expr: BinaryExpression, assumeTrue: boolean): Type {
+                assumeTrue = (expr.operatorToken.kind === SyntaxKind.EqualsEqualsEqualsToken) ? assumeTrue : !assumeTrue;
+                let lhs = skipParenthesizedNodes(expr.left);
+                const selectors: string[] = [];
+                while (lhs.kind !== SyntaxKind.Identifier) {
+                    switch (lhs.kind) {
+                        case SyntaxKind.PropertyAccessExpression:
+                            selectors.push((lhs as PropertyAccessExpression).name.text);
+                            // If a name doesn't resolve, bail
+                            if (selectors[selectors.length - 1] === undefined) {
+                                return type;
+                            }
+                            lhs = skipParenthesizedNodes((lhs as PropertyAccessExpression).expression);
+                            break;
+                        case SyntaxKind.Identifier:
+                            break;
+                        default:
+                            // Unhandled control flow construct, don't narrow
+                            return type;
+                    }
+                };
+                const identifier = lhs as Identifier;
+                if (getResolvedSymbol(identifier) !== symbol) {
+                    return type;
+                }
+                const rhsType = getTypeOfExpression(expr.right);
+                if (rhsType.flags & (TypeFlags.Undefined | TypeFlags.Null)) {
+                    return type;
+                }
+
+                // Descend into the type using the selectors we accumulated above and narrow any unions along the way
+                // If assumeTrue, we narrow by removing all types not compatible with the rhs type
+                // If not, we narrow only if the rhsType is a Value type (ie, StringLiteral) by removing all types compatible with that type (TODO)

```

Comment:
I think this comment should moved into a JSDoc coment for `narrowIntrospectively`